contracts/uniswap/UniswapV2Router02.sol

```diff
@@ -0,0 +1,270 @@
+pragma solidity ^0.6.0;
+
+import "@uniswap/lib/contracts/libraries/TransferHelper.sol";
+import "@uniswap/v2-core/contracts/interfaces/IUniswapV2Factory.sol";
+import "@uniswap/v2-core/contracts/interfaces/IUniswapV2Pair.sol";
+import "../interfaces/IERC20Extended.sol";
+import "../libraries/UniswapV2Library2.sol";
+import "./UniswapV2Router01.sol";
+
+contract UniswapV2Router02 is UniswapV2Router01 {
+    using SafeMath for uint256;
+
+    constructor(address _factory, address _weth) public UniswapV2Router01(_factory, _weth) {}
+
+    function addLiquiditySupportingFOTTokens(
+        address[2] calldata tokens,
+        uint256[2] calldata desiredAmounts,
+        uint256[2] calldata minAmounts,
+        bool[2] calldata areFOTTokens,
+        address to,
+        uint256 deadline
+    )
+        external
+        ensure(deadline)
+        returns (
+            uint256 amountA,
+            uint256 amountB,
+            uint256 liquidity
+        )
+    {
+        (amountA, amountB) = _addLiquidity(
+            tokens[0],
+            tokens[1],
+            desiredAmounts[0],
+            desiredAmounts[1],
+            minAmounts[0],
+            minAmounts[1]
+        );
+        address pair = UniswapV2Library.pairFor(factory, tokens[0], tokens[1]);
+        if (areFOTTokens[0]) {
+            IERC20Extended(tokens[0]).transferExactDestFrom(msg.sender, pair, amountA);
+        } else {
+            TransferHelper.safeTransferFrom(tokens[0], msg.sender, pair, amountA);
```

Comment:
https://github.com/Uniswap/uniswap-lib/blob/master/contracts/libraries/TransferHelper.sol#L20

The safety is at uniswap lib.

I don't want to change compare to uniswap code, so I use their libs instead of `SafeERC20`